internal/handlers/common/compare.go

```diff
@@ -188,15 +189,24 @@ func compare(docValue, filter any) compareResult {
 
 	switch docValue := docValue.(type) {
 	case *types.Document:
+		filter, isFilterDocument := filter.(*types.Document)
+		if isFilterDocument {
+			return compareDocuments(filter, docValue)
+		}
 		return notEqual
 
 	case *types.Array:
+		switch filter.(type) {
+		case *types.Array:
+			return compareArray(filter.(*types.Array), docValue)
+		}
+
 		for i := 0; i < docValue.Len(); i++ {
 			arrValue := must.NotFail(docValue.Get(i))
 
 			switch arrValue.(type) {
 			case *types.Document, *types.Array:
-				return notEqual
+				continue
```

Comment:
That looks very suspicious. `compare` checks if values are equal, not equal, less or greater. It does not checks conditions like "a is a subset of b". I'm not sure why we want to skip some values there.

Reply:
I guess we should apply the same logic as to array then with `compare` and `match` functions